cmd/gwi-api/api/api_integration_test.go

```diff
@@ -0,0 +1,133 @@
+package api_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api"
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api/stores/memorystore"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/labstack/echo"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/zap"
+)
+
+func TestAPI_Integration(t *testing.T) {
+	assert := assert.New(t)
+
+	must := func(err error) {
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Prepare the store.
+	store := memorystore.New()
+
+	// Prepare the API object.
+	a := api.New(zap.NewNop(), store, []byte("top-secret"))
+
+	// Prepare Echo.
+	e := echo.New()
+
+	// Create a new account.
+	c, _, rec := newRequest(t, e, "", bytes.NewBufferString(`{"email":"me@example.com","password":"secret"}`))
+	c.SetPath("/v1/volunteers/login")
+
+	must(a.PostLogin(c))
```

Comment:
This kind of test bypasses the router because you're calling the handler method directly. Do you have any other idea about how this test could be written? Explanation is fine, no need to code it up.

Reply:
You could simply invoke `ServeHTTP` on the `Echo` instance, that would include the router as well.

Or you can just start a testing HTTP server and bind the API to that, using https://godoc.org/net/http/httptest#Server, then test using regular HTTP requests.

Both of these then include the middleware as well, so JWT token management would need to be handled appropriately.